synapseclient/models/mixins/access_control.py

```diff
@@ -451,149 +637,1538 @@ async def _delete_current_entity_acl(
-                    )
+        entity_metadata = tree_structure["entity_metadata"]
+        children_map = tree_structure["children_map"]
+        root_entities = tree_structure["root_entities"]
 
-                    await folder.delete_permissions_async(
-                        include_self=should_delete_folder_acl,
-                        recursive=True,
-                        include_container_content=include_container_content,
-                        target_entity_types=target_entity_types,
-                        synapse_client=client,
-                    )
-                except Exception as e:
-                    client.logger.warning(
-                        f"Failed to delete ACL for folder {folder.id}: {str(e)}"
+        user_group_info_map = await self._fetch_user_group_info_from_tree(
+            tree_structure,
+            synapse_client or Synapse.get_client(synapse_client=synapse_client),
+        )
+
+        lines = []
+        lines.append("🔍 Legend:")
+        lines.append("  🔒 = No local ACL (inherits from parent)")
+        lines.append("  🔑 = Local ACL (custom permissions)")
+        lines.append("")
+
+        def get_acl_status_icon(entity_id: str) -> str:
+            """Get the appropriate ACL status icon for an entity."""
+            metadata = entity_metadata[entity_id]
+            acl = metadata.get("acl")
+            return (
+                "🔑"
+                if (acl and hasattr(acl, "acl_entries") and acl.acl_entries)
+                else "🔒"
+            )
+
+        def should_show_entity(entity_id: str) -> bool:
+            """Determine if an entity should be shown in the tree."""
+            metadata = entity_metadata[entity_id]
+            acl = metadata.get("acl")
+
+            if acl and hasattr(acl, "acl_entries") and acl.acl_entries:
+                return True
+
+            children = children_map.get(entity_id, [])
+            return any(should_show_entity(child_id) for child_id in children)
+
+        def format_entity_info(entity_id: str) -> str:
+            """Format entity information with ACL status icon."""
+            metadata = entity_metadata[entity_id]
+            name = metadata["name"]
+            entity_type = metadata["type"]
+            acl_icon = get_acl_status_icon(entity_id)
+            return f"{acl_icon} {name} ({entity_id}) [{entity_type}]"
+
+        def format_principal_info(principal_id: str, permissions: List[str]) -> str:
+            """Format principal information with permissions."""
+            permission_list = permissions if permissions else ["None"]
+            team_name = user_group_info_map.get(str(principal_id))
+
+            if team_name:
+                return (
+                    f"   └── {team_name} ({principal_id}): {', '.join(permission_list)}"
+                )
+            else:
+                return f"   └── Principal {principal_id}: {', '.join(permission_list)}"
+
+        def build_tree_recursive(
```

Comment:
Discussion: will these add a lot of methods to an Entity since this is a mixin?

Reply:
There are a bunch of `_*` functions added. These functions are defined within `_format_ascii_tree_async` and will not be exposed, since they are local functions to the scope of `_format_ascii_tree_async`